### PR TITLE
Games on Demand / Multi-file STFS (SVOD) Support

### PR DESCRIPTION
**Pull request by Wildenhaus**
Tested on several GoD container and XBLA Games (Burnout Revenge, Halo 3, Halo 3 ODST, JUJU, Duke Nukem 3D). Diff'ed various container files from the original source with no discrepancies. Handles both GDF and non-GDF formats with the same functions.

### CODE REVIEW

**Comment by Wildenhaus** [comment]
I just realized that this probably wont handle large files split up across multiple datafiles. Working on a fix, don't merge before it's appended.

**Comment by Wildenhaus** [comment]
Fixed device entries spanning multiple datafiles. Datafiles are split into ~162MB chunks. Tested with a few Halo 3 maps as well as Dead Space 2's bigfile0.viv (3GB). All files were diff'ed from the originals and no differences were found.

Aside from code review, this pull is in working state and ready for merge.

**Comment by x1nixmzeng** [comment]
Great work!

Could you double check the VFS dump tool works with your change?

https://github.com/xenia-project/xenia/tree/master/src/xenia/vfs

**Comment by Wildenhaus** [comment]
> Could you double check the VFS dump tool works with your change?

VFS Dump Tool is confirmed working. Top screenshot is dumping an nxeart STFS container, and bottom screenshot is a dump of a GoD Container, plus xenia running that dump.

vfsdump

If any issues arise, be sure to do a clean-rebuild. It was acting weird for me until I realized it was using an incomplete version of my changes to `stfs_container_device.cc`.

**Comment by gibbed** [comment]
1. Why were two ancient PRs pulled into this?
2. Please clean up your commit messages and squash unnecessary ones.

**Comment by Wildenhaus** [comment]
There are a few edge cases with offset calculation that are causing issues. Working on a more robust way to calculate offsets based on file layout.

**Comment by Wildenhaus** [comment]
- Rewrote SVOD offset calculation to no longer rely on known hard-coded addresses. Should now work with all cases. Added documentation for reference.
 - Added support for STFS Metadata v2

File: src/xenia/vfs/devices/stfs_container_device.cc [hunk 1]
```diff
@@ -617,6 +612,20 @@ bool StfsHeader::Read(const uint8_t* p) {
   title_thumbnail_image_size = xe::load_and_swap<uint32_t>(p + 0x1716);
   std::memcpy(thumbnail_image, p + 0x171A, 0x4000);
   std::memcpy(title_thumbnail_image, p + 0x571A, 0x4000);
+
+  // Metadata v2 Fields
+  std::memcpy(series_id, p + 0x3B1, 0x10);
```
Comment (x1nixmzeng) [inline src/xenia/vfs/devices/stfs_container_device.cc]: Should this be checking the metadata version is 2 (or greater) before reading these fields?

Comment (Wildenhaus) [reply src/xenia/vfs/devices/stfs_container_device.cc]: In the case of a game being Metadata v1, these offsets are unused, so they contain null data. Considering all of the fields added in Metadata v2 are irrelevant to emulation and will default to the same null data in `struct StfsHeader` if not read, I don't think it'll be an issue, but can be corrected if it's considered a code quality issue.

**Comment by Wildenhaus** [comment]
- Now supports all known layouts (EGDF, XSF, Single File)
- Major refactoring, code cleanup, and extensive documentation.
- Fixed code quality issue pointed out by @x1nixmzeng 

Tested both in-game and binary diff'ing VFS dump files. Works as it should for Digital Download, Installed, Arcade, and converted games (done by third-party tools, XSF).

File: src/xenia/vfs/devices/stfs_container_device.cc (hunk 1 above)
Comment (Wildenhaus) [reply src/xenia/vfs/devices/stfs_container_device.cc]: Conditionally parsing these fields as of the latest commit.

**Comment by Wildenhaus** [comment]
Corrected an issue where file entry ordinal calculation in games with a large amount of files (e.g. Mass Effect with 4508 files) would run into hash blocks.

**Comment by Ulquiorr4-Cifer** [comment]
Was this merged into DX12 or just master? 🤔

**Pull request merged**